applications/nrf_desktop/src/util/hid_reportq.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdint.h>
+#include <zephyr/sys/slist.h>
+#include <zephyr/kernel.h>
+
+#include "hid_reportq.h"
+#include "hid_report_desc.h"
+#include "hid_event.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(hid_reportq, CONFIG_DESKTOP_HID_REPORTQ_LOG_LEVEL);
+
+#define MAX_ENQUEUED_REPORTS CONFIG_DESKTOP_HID_REPORTQ_MAX_ENQUEUED_REPORTS
+
+struct enqueued_report {
+	sys_snode_t node;
+	struct hid_report_event *event;
+};
+
+struct counted_list {
+	sys_slist_t list;
+	size_t count;
+};
+
+struct hid_reportq {
+	struct counted_list reports[ARRAY_SIZE(input_reports)];
+	uint16_t enabled_report_idx_bm;
+	uint8_t last_sent_report_idx;
+	uint8_t report_max;
+	uint8_t report_cnt;
+	const void *sub_id;
+};
+
+static struct hid_reportq queues[CONFIG_DESKTOP_HID_REPORTQ_QUEUE_COUNT];
+
+/* Ensure that enabled_report_idx_bm can handle all of the report indexes. */
+BUILD_ASSERT(ARRAY_SIZE(input_reports) <= 16);
+
+static struct enqueued_report *get_enqueued_report(struct counted_list *l)
```

Comment:
I will call it `cl`, ok? Prefer short name for locally scoped variable. I also tried not to fully alias with a commonly used data structure name.

Reply:
Why avoiding longer names which are self describable? I still would prefer sth longer which self explains itself

Reply:
Maybe ask other members of team how they prefer to name variables

Reply:
Shorter names for locally scoped variables are nothing bad for me. They seem quite intuitive to use is use-cases like this one.
Can we keep `cl` here (@kapi-no, @alstrzebonski, @mkapala-nordic)? Let me know what do you think

Reply:
Personally I'd choose `cnt_list`, but I'm ok with `cl` too

Reply:
> Shorter names for locally scoped variables are nothing bad for me. They seem quite intuitive to use is use-cases like this one. Can we keep `cl` here (@kapi-no, @alstrzebonski, @mkapala-nordic)? Let me know what do you think

According to what I read in "Clean code" such short names with exceptions for iterators: "i", "j" ... shouldn't be used.
"q" is sth I can tolerate but honestly I think it is against the good practices. It cost almost nothing to have a longer name and can save a lot of time.

Reply:
I would use `list` as the param name to align with the slist API approach:

```
static inline bool sys_slist_is_empty(sys_slist_t *list);
```

Also, I would change the name of the structure to align with slist API `struct counted_list` -> `struct clist`

Reply:
> > Shorter names for locally scoped variables are nothing bad for me. They seem quite intuitive to use is use-cases like this one. Can we keep `cl` here (@kapi-no, @alstrzebonski, @mkapala-nordic)? Let me know what do you think
> 
> According to what I read in "Clean code" such short names with exceptions for iterators: "i", "j" ... shouldn't be used. "q" is sth I can tolerate but honestly I think it is against the good practices. It cost almost nothing to have a longer name and can save a lot of time.

I think that longer name is really not necessary for a function parameter name that is only used locally if the function is short. Still, I can call it `cnt_list` instead of `cl`.

Reply:
> I would use `list` as the param name to align with the slist API approach:
> 
> ```
> static inline bool sys_slist_is_empty(sys_slist_t *list);
> ```
> 
> Also, I would change the name of the structure to align with slist API `struct counted_list` -> `struct clist`

I prefer to keep the parameter name as `cnt_list`, because our struct contains a `list` field that is actually a single-linked list (so that we have `cnt_list->list` instead of `list->list` - it is a bit more readable). Using `struct clist` instead of `struct counted_list` would be against suggestion from @zycz (if we opt for using longer names, I would stick to `struct counted_list`).